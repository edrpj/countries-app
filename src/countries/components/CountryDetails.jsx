import { getCurrencyName } from "../helpers/getCurrencyName";
import { getLanguagesOfACountry } from "../helpers/getLanguagesOfACountry";
import { internationalNumberFormat } from "../helpers/internationalNumberFormat";

export const CountryDetails = ({
  name,
  nativeName,
  population,
  region,
  subregion,
  capital,
  topLevelDomain,
  currencies,
  languages,
  textColor,
}) => {
  return (
    <div className="row ms-0">
      <div className="row h-25 m-0 p-0">
        <div className="col ps-0 ps-lg-5 pt-4">
          <h2 className={`${textColor} font-extrabold`}>{name}</h2>
        </div>
      </div>

      <div className="row h-75 m-0 p-0 pt-3">
        <div className="col-12 col-sm-6 ps-0 ps-lg-5">
          <p aria-label="paragraph" className={`${textColor} mb-2`}>
            <span className="text-base font-semibold">Native Name: </span>{" "}
            {nativeName ?? name}
          </p>
          <p aria-label="paragraph" className={`${textColor} mb-2`}>
            <span className="text-base font-semibold">Population: </span>{" "}
            {internationalNumberFormat.format(population)}
          </p>
          <p aria-label="paragraph" className={`${textColor} mb-2`}>
            <span className="text-base font-semibold">Region: </span> {region}
          </p>
          <p aria-label="paragraph" className={`${textColor} mb-2`}>
            <span className="text-base font-semibold">Sub Region: </span>{" "}
            {subregion}
          </p>
          <p aria-label="paragraph" className={`${textColor} mb-2`}>
            <span className="text-base font-semibold">Capital: </span> {capital}
          </p>
        </div>

        <div className="col-12 col-sm-6 ps-0 ps-sm-4 ps-md-2 mt-4 mt-sm-0">
          <p aria-label="paragraph" className={`${textColor} mb-2`}>
            <span className="text-base font-semibold">Top Level Domain: </span>{" "}
            {topLevelDomain}
          </p>
          <p aria-label="paragraph" className={`${textColor} mb-2`}>
            <span className="text-base font-semibold">Currencies: </span>{" "}
            {getCurrencyName(currencies)}
          </p>
          <p aria-label="paragraph" className={`${textColor} mb-2`}>
            <span className="text-base font-semibold">Languages: </span>{" "}
            {getLanguagesOfACountry(languages)}
          </p>
        </div>
      </div>
    </div>
  );
};
