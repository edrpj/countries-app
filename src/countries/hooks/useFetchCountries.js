import { useEffect, useState } from "react";
import { getCountries } from "../services/getCountries";

export const useFetchCountries = (query) => {
  const [countries, setCountries] = useState([]);
  const [isLoading, setIsLoading] = useState(true);

  const fetchCountries = async () => {
    try {
      const countriesList = await getCountries(query);
      setCountries(countriesList);
      setIsLoading(false);
    } catch (error) {
      setCountries([]);
      setIsLoading(false);
    }
  };

  useEffect(() => {
    fetchCountries();
  }, [query]);

  return {
    countries,
    isLoading,
  };
};
