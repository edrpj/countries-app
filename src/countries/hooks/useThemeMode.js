import { useContext } from "react";

import { ThemeContext } from "../../theme/ThemeContext";
import { themeType } from "../../theme/themeType";
import { colorType } from "../types/colorType";

const {
  bgLightMode,
  bgElementsDarkMode,
  bgContainerDarkMode,
  bgBtnLightMode,
  textDarkMode,
  textLightMode,
  textInputDarkMode,
  textInputLightMode,
} = colorType;

export const useThemeMode = () => {
  const themeContextValue = useContext(ThemeContext);
  const { theme, onChangeTheme } = themeContextValue;

  const themeMode = localStorage.getItem("mode");
  const isLight = themeMode === themeType.light;

  const bgColorContainer = isLight ? bgLightMode : bgContainerDarkMode;
  const bgColorElements = isLight ? bgLightMode : bgElementsDarkMode;
  const bgColorBtn = isLight ? bgBtnLightMode : bgElementsDarkMode;
  const textColor = isLight ? textLightMode : textDarkMode;
  const textInputColor = isLight ? textInputLightMode : textInputDarkMode;

  return {
    bgColorContainer,
    bgColorElements,
    bgColorBtn,
    textColor,
    textInputColor,
    isLight,
    onChangeTheme,
    theme,
  };
};
