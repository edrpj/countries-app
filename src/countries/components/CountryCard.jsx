import { useNavigate } from "react-router-dom";
import { internationalNumberFormat } from "../helpers/internationalNumberFormat";

import { useThemeMode } from "../hooks/useThemeMode";

export const CountryCard = ({
  id,
  name,
  capital,
  population,
  region,
  imgUrl,
}) => {
  const navigate = useNavigate();
  const { bgColorElements, textColor } = useThemeMode();

  const onNavigateToACard = () => {
    navigate(`/country/${id}`);
  };

  return (
    <div className="col">
      <div
        aria-label="card"
        className={`card shadow cursor-pointer ${bgColorElements}`}
        onClick={onNavigateToACard}
      >
        <img
          src={imgUrl}
          className="card-img-top w-100 object-cover h-img-card"
          alt={name}
        />

        <div aria-label="information" className={`card-body ${textColor}`}>
          <h5 className="card-title text-base font-extrabold mb-3">{name}</h5>
          <p className="card-text mb-0">
            <span className="text-sm font-extrabold">Population: </span>{" "}
            {internationalNumberFormat.format(population)}
          </p>
          <p className="card-text mb-0">
            <span className="text-sm font-extrabold">Region: </span> {region}
          </p>
          <p className="card-text mb-3">
            <span className="text-sm font-extrabold">Capital: </span> {capital}
          </p>
        </div>
      </div>
    </div>
  );
};
