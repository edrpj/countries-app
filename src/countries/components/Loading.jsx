import { useThemeMode } from "../hooks/useThemeMode";

export const Loading = () => {
  const { isLight, bgColorContainer } = useThemeMode();

  return (
    <div
      className={`w-100 h-100 m-0-auto d-flex justify-content-center flex-1 ${bgColorContainer}`}
    >
      <div
        aria-label="spinner"
        className={`spinner-border ${isLight ? "text-dark" : "text-warning"}`}
        style={{ width: "5rem", height: "5rem" }}
        role="status"
      >
        <span className="visually-hidden">Loading...</span>
      </div>
    </div>
  );
};
