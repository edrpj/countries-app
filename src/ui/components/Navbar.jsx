import { Link } from "react-router-dom";

import { useThemeMode } from "../../countries/hooks/useThemeMode";

export const Navbar = () => {
  const { bgColorElements, textColor, onChangeTheme, theme } = useThemeMode();

  return (
    <nav className={`navbar shadow-sm py-3 ${bgColorElements}`}>
      <div className="container-fluid px-0 w-90 m-auto">
        <Link
          aria-label="logo"
          className={`navbar-brand font-extrabold ${textColor}`}
          to="countries"
        >
          Where in the world?
        </Link>
        <div
          aria-label="changeThemeDiv"
          className="d-flex align-items-center cursor-pointer"
          onClick={() => onChangeTheme(theme)}
        >
          <i
            aria-label="iTag"
            className={`fa-regular fa-moon d-flex align-items-center ${textColor}`}
          ></i>
          <p
            aria-label="paragraph"
            className={`${textColor} font-semibold ms-2 m-0-auto`}
          >
            Dark Mode
          </p>
        </div>
      </div>
    </nav>
  );
};
