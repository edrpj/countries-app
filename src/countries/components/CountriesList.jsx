import { useFetchCountries } from "../hooks/useFetchCountries";
import { CountryCard } from "./CountryCard";
import { Loading } from "./Loading";

export const CountriesList = ({ query }) => {
  const { countries, isLoading } = useFetchCountries(query);

  if (isLoading) return <Loading />;

  return (
    <div className="row row-cols-1 row-cols-sm-2 row-cols-md-3 row-cols-xl-4 g-5">
      {countries.map((country) => {
        const { cca3, name, population, region, capital, flags } = country;

        return (
          <CountryCard
            key={cca3}
            id={cca3}
            name={name.common}
            capital={capital}
            population={population}
            region={region}
            imgUrl={flags.svg}
          />
        );
      })}
    </div>
  );
};
