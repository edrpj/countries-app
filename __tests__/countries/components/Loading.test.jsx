import { render, screen } from "@testing-library/react";
import { Loading } from "../../../src/countries/components/Loading";
import { useThemeMode } from "../../../src/countries/hooks/useThemeMode";

jest.mock("../../../src/countries/hooks/useThemeMode");

describe("Testing <Loading />", () => {
  it("should do match with the snapshot", () => {
    useThemeMode.mockReturnValue({
      isLight: true,
      bgColorContainer: "bg-white",
    });

    const { container } = render(<Loading />);
    expect(container).toMatchSnapshot();
  });

  it("should use the correct text color when the app is in light mode", () => {
    const useThemeModeData = {
      isLight: true,
      bgColorContainer: "bg-white",
    };
    useThemeMode.mockReturnValue(useThemeModeData);

    render(<Loading />);
    const spinner = screen.getByLabelText("spinner");
    expect(spinner.className).toContain("text-dark");
  });

  it("should use the correct text color when the app is in dark mode", () => {
    const useThemeModeData = {
      isLight: false,
      bgColorContainer: "bg-white",
    };
    useThemeMode.mockReturnValue(useThemeModeData);

    render(<Loading />);
    const spinner = screen.getByLabelText("spinner");
    expect(spinner.className).toContain("text-warning");
  });
});
