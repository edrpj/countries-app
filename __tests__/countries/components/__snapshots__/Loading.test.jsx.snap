// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <Loading /> should do match with the snapshot 1`] = `
<div>
  <div
    class="w-100 h-100 m-0-auto d-flex justify-content-center flex-1 bg-white"
  >
    <div
      aria-label="spinner"
      class="spinner-border text-dark"
      role="status"
      style="width: 5rem; height: 5rem;"
    >
      <span
        class="visually-hidden"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;
